File: SPOJ/EPALIN.cpp
#include<iostream>
#include<string>
using namespace std;

int main()
{
    string s;
    while(getline(cin,s))
    {
      int n = s.length();
      int MOD = 4084061;
      int prod=1,k = 26;
      int hashs=0,hasht=0,count=0,max=0;
      for(int i=n-1;i>-1;i--)
      {
        hashs*=k;hashs+=s[i]-97;
        hashs%=MOD;
        hasht+=(s[i]-97)*prod;
        hasht%=MOD;
        if(hashs==hasht)max=n-i;
        prod*=k;prod%=MOD;
      }
      count = max;
      string a,b,c,d;
      a = s.substr(0,n-count);
      b = s.substr(n-count,count+1);
      reverse(a.begin(),a.end());
      c = a;
      reverse(a.begin(),a.end());
      d = a+b+c;
      cout<<d<<endl;
    }
}

File: Competitive-Programming/SPOJ/EPALIN.py
while True:
    try: s = raw_input()
    except:break;exit()
    n = len(s)
    d = {}
    MOD = 4084061
    prod = 1
    k  = 26
    hashs,hasht,count = 0,0,0
    max = 0
    for i in range(n-1,-1,-1):
        hashs*=k
        hashs+=ord(s[i])-97
        hashs%=MOD
        hasht+=(ord(s[i])-97)*prod
        hasht%=MOD
        if hashs==hasht:max = n-i
        prod*=k
    count  = max
    print s[:n-count]+s[n-count:]+s[:n-count][::-1]
